lib/sshutils/sftp/sftp.go

```diff
@@ -0,0 +1,472 @@
+	Open(ctx context.Context, path string) (io.ReadCloser, error)
+	// Create creates a new file
+	Create(ctx context.Context, path string, mode os.FileMode) (io.WriteCloser, error)
+	// MkDir creates a directory
+	// sftp.Client.Mkdir does not take an os.FileMode, so this can't either
+	Mkdir(ctx context.Context, path string, mode os.FileMode) error
+	// Chmod sets file permissions
+	Chmod(ctx context.Context, path string, mode os.FileMode) error
+	// Chtimes sets file access and modification time
+	Chtimes(ctx context.Context, path string, atime, mtime time.Time) error
+}
+
+// CreateUploadConfig returns a Config ready to upload files
+func CreateUploadConfig(src []string, dst string, opts Options) (*Config, error) {
+	for _, srcPath := range src {
+		if srcPath == "" {
+			return nil, trace.BadParameter("source path is empty")
+		}
+	}
+	if dst == "" {
+		return nil, trace.BadParameter("destination path is empty")
+	}
+
+	c := &Config{
+		srcPaths: src,
+		dstPath:  dst,
+		srcFS:    &localFS{},
+		dstFS:    &remoteFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c, nil
+}
+
+// CreateDownloadConfig returns a Config ready to download files
+func CreateDownloadConfig(src, dst string, opts Options) (*Config, error) {
+	if src == "" {
+		return nil, trace.BadParameter("source path is empty")
+	}
+	if dst == "" {
+		return nil, trace.BadParameter("destination path is empty")
+	}
+
+	c := &Config{
+		srcPaths: []string{src},
+		dstPath:  dst,
+		srcFS:    &remoteFS{},
+		dstFS:    &localFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c, nil
+}
+
+// setDefaults sets default values
+func (c *Config) setDefaults() {
+	logger := c.Log
+	if logger == nil {
+		logger = log.StandardLogger()
+	}
+	c.Log = logger.WithFields(log.Fields{
+		trace.Component: "SFTP",
+		trace.ComponentFields: log.Fields{
+			"SrcPaths":      c.srcPaths,
+			"DstPath":       c.dstPath,
+			"Recursive":     c.opts.Recursive,
+			"PreserveAttrs": c.opts.PreserveAttrs,
+		},
+	})
+}
+
+// TransferFiles transfers files from the configured source paths to the
+// configured destination path over SFTP
+func (c *Config) TransferFiles(ctx context.Context, sshClient *ssh.Client) error {
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := c.initFS(ctx, sshClient, sftpClient); err != nil {
+		return trace.Wrap(err)
+	}
+
+	transferErr := c.transfer(ctx)
+	closeErr := sftpClient.Close()
+	if transferErr != nil {
+		return trace.Wrap(transferErr)
+	}
+
+	return trace.Wrap(closeErr)
+}
+
+// initFS ensures the source and destination filesystems are ready to transfer
+func (c *Config) initFS(ctx context.Context, sshClient *ssh.Client, client *sftp.Client) error {
+	var haveRemoteFS bool
+
+	srcFS, srcOK := c.srcFS.(*remoteFS)
+	if srcOK {
+		srcFS.c = client
+		haveRemoteFS = true
+	}
+	dstFS, dstOK := c.dstFS.(*remoteFS)
+	if dstOK {
+		dstFS.c = client
+		haveRemoteFS = true
+	}
+
+	if haveRemoteFS {
```

Comment:
nit: You can revert the condition to remove one nested if.
```suggestion
	if !haveRemoteFS {
		return nil
	}
	...
```

Reply:
Done